x/zoneconcierge/keeper/hooks.go

```diff
@@ -0,0 +1,65 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	checkpointingtypes "github.com/babylonlabs-io/babylon/x/checkpointing/types"
+	epochingtypes "github.com/babylonlabs-io/babylon/x/epoching/types"
+	"github.com/babylonlabs-io/babylon/x/zoneconcierge/types"
+)
+
+type Hooks struct {
+	k Keeper
+}
+
+// ensures Hooks implements ClientHooks interfaces
+var _ checkpointingtypes.CheckpointingHooks = Hooks{}
+var _ epochingtypes.EpochingHooks = Hooks{}
+
+func (k Keeper) Hooks() Hooks { return Hooks{k} }
+
+func (h Hooks) AfterEpochEnds(ctx context.Context, epoch uint64) {
+	// upon an epoch has ended, index the current chain info for each CZ
+	// TODO: do this together when epoch is sealed?
+	for _, consumerID := range h.k.GetAllConsumerIDs(ctx) {
+		h.k.recordEpochChainInfo(ctx, consumerID, epoch)
+	}
+}
+
+func (h Hooks) AfterRawCheckpointSealed(ctx context.Context, epoch uint64) error {
+	// upon a raw checkpoint is sealed, index the current chain info for each consumer,
+	// and generate/save the proof that the epoch is sealed
+	h.k.recordEpochChainInfoProofs(ctx, epoch)
+	h.k.recordSealedEpochProof(ctx, epoch)
+	return nil
+}
+
+// AfterRawCheckpointFinalized is triggered upon an epoch has been finalised
+func (h Hooks) AfterRawCheckpointFinalized(ctx context.Context, epoch uint64) error {
+	headersToBroadcast := h.k.getHeadersToBroadcast(ctx)
+
+	// send BTC timestamp to all open channels with ZoneConcierge
+	h.k.BroadcastBTCTimestamps(ctx, epoch, headersToBroadcast)
+
+	// Update the last broadcasted segment
+	h.k.setLastSentSegment(ctx, &types.BTCChainSegment{
+		BtcHeaders: headersToBroadcast,
+	})
```

Comment:
I did not understood why we keep the last sent segment

Reply:
Babylon needs to send the *new* BTC headers during the last finalised epoch to consumer chains. This means

- Babylon needs to at least remember the last BTC header sent in the last finalisd epoch
- However, BTC chain can get reorg'ed. So Babylon needs to remember at least the last k BTC headers